### PR TITLE
Various chat improvements

### LINKED ISSUE #675: video preview broken
Videos do not show in chat screen, instead a weird d-style box is shown:

Screenshot_20230627-160613

### LINKED ISSUE #861: Gallery not being accessed while uploading image in chat room.

### LINKED ISSUE #1115: Open and view files and/or images/videos
when clicking on an image within the chat, nothing happens. Instead a large zoom in of the image should be made available so the user can look at it more closely. Same for videos. For other files, the user should be prompted if they want to download and open it and if agreed do so ;) .

### LINKED ISSUE #1117: Unable to upload on the chat
Error on uploading:
https://github.com/acterglobal/a3/assets/51526480/a494fbae-77b4-4543-8081-90b70bbda5ba

Unable to see video message:
https://github.com/acterglobal/a3/assets/51526480/43c37964-117d-4133-95a9-b236bd136b19

### PR DESCRIPTION
**Pull request by kumarpalsinh25**
This PR includes various improvement in chat user experience as below.

1. Chat Attachments:
Now user have separate options for chat attachment like Camera, Image, Video and File

2. Video Chat message support:
Now have can share video chat message and also view it on chat area

3. Full screen chat image/video support:
Now user can see chat image/video in full screen for better viewing experience.

4. Share image/video:
User can share image/video from full screen video by clicking on share button

5. General improvements

- - - - -
- fixes #675 
- fixes #861 
- ref #1115 
- fixes #1117 
- - - - -

https://github.com/acterglobal/a3/assets/51526480/ce9f470e-3c63-4fa6-a3d3-b9ad03459ebb

### CODE REVIEW

File: app/lib/common/providers/chat_providers.dart [hunk 1]
```diff
@@ -125,3 +129,49 @@ final memberProvider =
...
+  var imageFile = File('${tempDir.path}/image-$messageId.jpg');
+  var isImageAvailable = await imageFile.exists();
+
+  if (!isImageAvailable) {
+    if (convo != null) {
+      //If video file is not available on local store then save it
+      AsyncValue<Uint8List> videoData = AsyncValue.data(
+        await convo.imageBinary(messageId).then((value) => value.asTypedList()),
+      );
+      if (videoData.asData != null) {
```
Comment (gtalha07) [inline app/lib/common/providers/chat_providers.dart:147]: We should probably handle error case here like some error string specifying that media can't be opened or unavailable.

File: app/lib/common/providers/chat_providers.dart [hunk 2]
```diff
@@ -125,3 +129,49 @@ final memberProvider =
...
+  var videoFile = File('${tempDir.path}/video-$messageId.mp4');
+  var isVideoAvailable = await videoFile.exists();
+
+  if (!isVideoAvailable) {
+    if (convo != null) {
+      //If video file is not available on local store then save it
+      AsyncValue<Uint8List> videoData = AsyncValue.data(
+        await convo.videoBinary(messageId).then((value) => value.asTypedList()),
+      );
+      if (videoData.asData != null) {
```
Comment (gtalha07) [inline app/lib/common/providers/chat_providers.dart:170]: Same as above.

**Review by gtalha07 — APPROVED** [review]
This embarks a significant improvement for chats in general. Very well put and excited to have this merged and get in use.

File: app/lib/features/chat/widgets/chat_attachment_options.dart [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import 'package:acter/common/themes/app_theme.dart';
+import 'package:atlas_icons/atlas_icons.dart';
+import 'package:flutter/material.dart';
+
+class ChatAttachmentOptions extends StatelessWidget {
+  final VoidCallback? onTapCamera;
```
[the remark this answers is not in the document]
Comment (kumarpalsinh25) [reply app/lib/features/chat/widgets/chat_attachment_options.dart:6]: Sorry but not getting you here. 
Camera button is fully functional for mobile platform (Check video in PR description). 

Let me know if any specific thing about the camera feature is missing.

File: app/lib/features/chat/widgets/chat_attachment_options.dart [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      physics: const NeverScrollableScrollPhysics(),
+      children: [
+        if (!isDesktop)
+          ListTile(
+            onTap: () {
+              Navigator.of(context).pop();
+              if (onTapCamera != null) onTapCamera!();
+            },
+            leading: Icon(Atlas.camera, color: iconColor),
+            title: const Text('Camera'),
```
[the remark this answers is not in the document]
Comment (kumarpalsinh25) [reply app/lib/features/chat/widgets/chat_attachment_options.dart:33]: Same as above comment.
https://github.com/acterglobal/a3/pull/1197#discussion_r1419959786

File: app/lib/features/chat/widgets/chat_attachment_options.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (gtalha07) [reply app/lib/features/chat/widgets/chat_attachment_options.dart:6]: Oh it does? That's awesome. Ignore then ;)

**Pull request merged**